### PR TITLE
fix #15959

### LINKED ISSUE #15959: in generics, params and return can't call procs in for their type declaration; gives Error: type mismatch
### Example 1
`lent type(expr)` gives Error: type mismatch
```nim
proc byLent2[T](a: T): lent type(a[0]) = a[0] # Error: type mismatch: got <T, int literal(0)>
# proc byLent2[T](a: T): lent typeof(a[0]) = a[0] # ditto
# proc byLent2[T](a: T): lent[type(a[0])] = a[0] # Error: no generic parameters allowed for lent
```

### Current Output
```
Error: type mismatch: got <T, int literal(0)>
```

### Expected Output
works

### [EDIT] Example 2
```nim
when true: # templates work; and `typeof` special case also works:
  template fn(a): untyped = a
  proc my[T](a: T): typeof(fn(a)) = discard

when true: # procs don't
  proc fn[U](a: U): auto = a
  proc my[T](a: T): typeof(fn(a)) = discard

when true: # ditto with params
  proc fn[U](a: U): auto = a
  proc my[T](a: T, b: typeof(fn(a))) = discard

when true: # non-generic procs don't have this problem
  proc fn[U](a: U): auto = a
  proc my(a: int): typeof(fn(a)) = discard
```

### Additional Information
* devel 1.5.1 c39fa0d49535da7f803d9972abdb15d09eed92a6
* might be related to static early symbol resolution (eg https://github.com/nim-lang/Nim/issues/8677)
* added disabled test case in https://github.com/nim-lang/Nim/pull/15960

### notes
default values for optionals can call procs, unrestricted:
```nim
when true:
  proc fn[U](a: U): auto = a
  proc my[T](a: T, b = typeof(fn(a)).default) = discard
```

### PR DESCRIPTION
**Pull request by cooldome**
Please review the approach. My knowledge of generics is still lacking.
I have added `typeFromExpr` for subscript operator, but I was getting cannot Eval error from the vm in `replaceTypeVarsTAux`.

Because of this I have patched `replaceTypeVarsTAux` to call `semExpr`  instead of `semConstExpr` and use `semConstExpr` only if needed. This step I am not sure about.

### CODE REVIEW

**Comment by timotheecour** [comment]
this doesn't `fix #15959`, only a particular case. Commit + PR should be changed to `refs #15959`, many examples there still don't work.
But I don't think special casing subscript operator is a good idea, a general fix is needed otherwise you'll end up with lots of complexity to handle the other cases similarly.

`typeFromExpr` is indeed the proper fix, but it should be applied in the general case, perhaps as a catch-all.

**Comment by cooldome** [comment]
I have tried catch all `typeFromExpr`, it creates a lot more problems than fixes.
Special casing for subscrit is required because `sem` has special case for it, if `a[0]` was a generic function call than special casing would not be required. BracketExpr remains special expression, so its type.

**Comment by cooldome** [comment]
Test failure is unrelated

**Comment by timotheecour** [comment]
either way, this still holds:
> this doesn't fix #15959, only a particular case. Commit + PR should be changed to refs #15959, many examples there still don't work.

File: compiler/semexprs.nim [hunk 1]
```diff
@@ -1513,6 +1513,9 @@ proc semSubscript(c: PContext, n: PNode, flags: TExprFlags): PNode =
       result = n
     else:
       result = nil
+  of tyGenericParam:
+    result = n
+    result.typ = makeTypeFromExpr(c, n.copyTree)
```
Comment (Araq) [inline compiler/semexprs.nim:1518]: That seems to be a bad idea, can't we add a better proc/syntax/interface to ``tyGenericParam``. The ``[]`` is very overloaded already, seems to be recipe for desaster.

Comment (cooldome) [reply compiler/semexprs.nim:1518]: Not sure I understand. If `[]` was a proc defined in system then this change wouldn't be required, but nkBracketExpr is still the special beast that is handled by its own rules. These rules don't handle `tyGenericParam`, hence the fix.

Another solution would be to define all missing procs in system.nim
```nim
proc `[]`[T: tuple](x: T, i: int): T {.magic: "XXX".}
```
Possibly others and then remove most of `semSubscript` and rely on operator overloading. 
Not sure this change will be easy and go smoothly though.

Comment (timotheecour) [reply compiler/semexprs.nim:1518]: `[]` special rules indeed cause a lot of issues (eg https://github.com/nim-lang/Nim/issues/15911) and should be replaced by proper overloading (of possibly magic procs)

**Pull request closed without merging**